From 65d107823899a2a0b992d55de623b703ac2f56e6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jan=20Kl=C3=B6tzke?= <jan@kloetzke.net>
Date: Sun, 11 Sep 2022 20:29:03 +0200
Subject: [PATCH 06/17] winex11.drv: optimize repeated ClipCursor calls
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

At least Doom 64 calls ClipCursor() with identical parameters
repeatedly, which seems to cause a considerable overhead. Together with
high polling rate mouse input this causes the game to almost freeze
while the mouse is being moved. So if the clipping did not change we can
bail out early because it will not cause any observable difference
anyway.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=46976
Signed-off-by: Jan Klötzke <jan@kloetzke.net>
---
 dlls/winex11.drv/mouse.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index 6af78211d74..ad10d30341e 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -471,6 +471,10 @@ static BOOL grab_clipping_window( const RECT *clip )
     if (!is_current_process_focused()) return TRUE;
 
     if (!data) return FALSE;
+
+    if (clipping_cursor && EqualRect( clip, &clip_rect ))
+        return TRUE;
+
     if (!(clip_window = init_clip_window())) return TRUE;
 
     if (keyboard_grabbed)
-- 
2.55.0

